src/Microsoft.Windows.CsWin32/templates/HRESULT.cs

```diff
@@ -61,7 +61,9 @@ internal HRESULT ThrowOnFailure(IntPtr errorInfo = default)
         return this;
     }
 
-    public override string ToString() => this.Value.ToString();
+	internal static HRESULT HRESULT_FROM_WIN32(WIN32_ERROR error) => new(((int)error & 0x0000FFFF) | unchecked((int)0x80070000));
```

Comment:
Are we sure the conversion is this simple? I'm looking at [this doc](https://docs.microsoft.com/en-us/windows/win32/api/winerror/nf-winerror-hresult_from_win32#remarks) and it seems more complicated than this.

[This C# HResult struct](https://github.com/dotnet/pinvoke/blob/main/src/Windows.Core/HResult.cs) may be a resource as well.

Reply:
Should I be adding via generator is the dependency is added when importing HRESULT?

Reply:
@AArnot, I guessing winforms did what they did because `(FACILITY_WIN32 << 16) | 0x80000000)` effectively is `unchecked((int)0x80070000)`? I don't know for sure, but I think its the bitshift on the const?

Reply:
> Should I be adding via generator so the dependency is added when importing HRESULT?

Yes, we absolutely should ensure the referenced types are generated. We can probably combine the template simplicity with a special step in the generator that 'knows' to add the types referenced in the template to try to capture the best of both worlds.

The difference in the expression that I was focusing on is that negative WIN32_ERROR is supposed to be taken as-is. Only positive values get this special bit-wise treatment.

Reply:
> Yes, we absolutely should ensure the referenced types are generated. We can probably combine the template simplicity with a special step in the generator that 'knows' to add the types referenced in the template to try to capture the best of both worlds.

I have no idea how to implement that.

> The difference in the expression that I was focusing on is that negative WIN32_ERROR is supposed to be taken as-is. Only positive values get this special bit-wise treatment.

The negative test was added in windows xp. Nothing on why though. I imagine the negative values were added around then and maybe they were managed better in `WIN32_ERROR`? Who knows.
https://matthewvaneerde.wordpress.com/2007/12/28/the-evolution-of-hresult_from_win32/

Raymond Chen also wrote an article on doing the reverse mapping which might be useful.
https://devblogs.microsoft.com/oldnewthing/20061103-07/?p=29133

Reply:
@russkie sorry to bother you. But do you know why there is a negative check in `HRESULT_FROM_WIN32` and why there isn't a bit shift on it?

Reply:
Regardless of the reason the negativity check exists, for auditing purposes I'd prefer that we kept as close to the same syntax as the header files as possible. 
I can help with the testing.

Reply:
Is there anywhere else in the code where we detect dependencies and then include them?

Reply:
@ecalalang I am afraid I don't know. I think you know about this speific area way more than me at this point :)

Reply:
Thanks, do you know someone that might know?

Reply:
May be try nerd sniping @oldnewthing? :D 

My first point of contact for such things is generally @JeremyKuhne, but he's OOF right now.

Reply:
@elachlan, @RussKie presumably the negative check there is to deal with things that are already `HRESULT`s or possibly filter out other ["custom" codes](https://docs.microsoft.com/windows/win32/debug/system-error-codes#third-party-error-codes) that might be set on last error, but I'm not 100% sure. In any case, something that was negative wouldn't make much sense wrapping in an `HRESULT` and claiming it was a Win32 error. I'd just follow along with the actual implementation.